NDTensors/src/dense/densetensor.jl

```diff
@@ -93,6 +93,12 @@ end
 @propagate_inbounds function setindex!(
   T::DenseTensor{<:Number}, x::Number, I::Vararg{Integer}
 )
+  if is_unallocated_zeros(T)
+    ElT = promote_type(eltype(data(T)), typeof(x))
+    d = similartype(alloctype(data(T)), ElT)(undef, dim(to_shape(typeof(data(T)), inds(T))))
+    fill!(d, zero(ElT))
+    T = Tensor(Dense(d), inds(T))
+  end
```

Comment:
Also I'm sure there are many places we will want to allocate a `Zeros` or a `TensorStorage` or `Tensor` wrapping `Zeros` data, probably good to define a generic function (maybe called `allocate`) which allocates `Zeros` data based on `alloctype`, and can do that for `TensorStorage` and `Tensor` types. Seems like `adapt` would be a good way to implement that generically for types like `TensorStorage` and `Tensor`, maybe through a new `AllocAdaptor`.

Reply:
Yeah I thought I might have put this operation in the wrong location. I can fix that. I definitely agree, I already found that I need to implement an allocate (and type promotion) in the blocksparse setindex!! code as well so it would make sense to have a easily reusable function to simplify the process!